asynciterator.ts

```diff
@@ -777,9 +784,121 @@ export class IntegerIterator extends AsyncIterator<number> {
+ */
+export type MapFunction<S, D = S, I extends AsyncIterator<D> = AsyncIterator<D>> =
+  (item: S, iterator: I) => D | null;
+
+/**
+  An iterator that calls a synchronous mapping function
+  on every item from its source iterator.
+  @extends module:asynciterator.AsyncIterator
+*/
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  protected _source: AsyncIterator<S>;
+  private readonly _destroySource: boolean;
+  private readonly _mappings: MapFunction<any, any, MappingIterator<S, D>>[];
+  private readonly _mappingRoot: InternalSource<any>;
+
+  // This is wrong: readable should be set by listening to source events
+  get readable() {
+    return (this._state < CLOSED) && this._source.readable;
+  }
+
+  /**
+   * Applies the given mapping to the source iterator.
+   */
+  constructor(
+    source: AsyncIterator<S>,
+    mapping?: MapFunction<S, D, MappingIterator<S, D>>,
+    options?: SourcedIteratorOptions,
+  );
+
+  /**
+   * Applies the given list of mappings to the mapping root.
+   *
+   * This is an optimization for
+   *     root.map(f1).map(f2).map(f3)
+   * where the combined mapping x => f3(f2(f1(x)))
+   * is applied to root rather than to the intermediate sources.
+   */
+  constructor(
+    source: AsyncIterator<S>,
+    mappings: MapFunction<any, any, MappingIterator<S, D>>[],
+    mappingRoot: AsyncIterator<any>,
+    options?: SourcedIteratorOptions,
+  );
+
+  constructor(
+    source: AsyncIterator<S>,
+    mappings: MapFunction<S, D, MappingIterator<S, D>> |
+              MapFunction<any, any, MappingIterator<S, D>>[] = [],
+    mappingRoot?: AsyncIterator<any> | SourcedIteratorOptions,
+    options: SourcedIteratorOptions = {},
+  ) {
+    super();
+    // Resolve optional parameters
+    if (!isEventEmitter(mappingRoot)) {
+      if (mappingRoot)
+        options = mappingRoot;
+      mappingRoot = source;
+    }
+    this._source = source;
+    this._mappings = isFunction(mappings) ? [mappings] : mappings;
+    this._mappingRoot = mappingRoot as InternalSource<any>;
+    this._destroySource = options.destroySource !== false;
+
+    if (mappingRoot.done) {
+      this.close();
+    }
+    else {
+      _validateSource(mappingRoot);
+      this._mappingRoot._destination = this;
+      this._mappingRoot.on('end', destinationClose);
+      this._mappingRoot.on('error', destinationEmitError);
+      this._mappingRoot.on('readable', destinationEmitReadable);
+    }
+  }
+
+  read(): D | null {
+    let mapped : any = null;
+    while (mapped === null && (mapped = this._source.read()) !== null) {
+      for (let i = 0; i < this._mappings.length; i++) {
+        mapped = this._mappings[i](mapped, this);
+        if (mapped === null)
+          break;
+      }
+    }
+    return mapped;
+  }
+
+  map<K>(map: (item: D, it: AsyncIterator<any>) => K | null, self?: any): AsyncIterator<K> {
+    return new MappingIterator<S, K>(this._source, [...this._mappings, bind(map, self)], this);
```

Comment:
Something is off here.

`MappingIterator` (overload 2) takes as constructor parameters the direct source (should be `this`), the mappings, and the mapping root (should be `this._mappingRoot`). Yet the tests do not catch this problem.